src/ProductManifold.jl

```diff
@@ -372,22 +372,49 @@ function default_vector_transport_method(M::ProductManifold)
     return ProductVectorTransport(map(default_vector_transport_method, M.manifolds)...)
 end
 
-@doc raw"""
-    distance(M::ProductManifold, p, q)
-
-Compute the distance between two points `p` and `q` on the [`ProductManifold`](@ref) `M`, which is
-the 2-norm of the elementwise distances on the internal manifolds that build `M`.
-"""
-function distance(M::ProductManifold, p, q)
-    return sqrt(
-        sum(
-            map(
-                distance,
-                M.manifolds,
-                submanifold_components(M, p),
-                submanifold_components(M, q),
-            ) .^ 2,
+_doc_distance_prod = """
+    distance(M::ProductManifold, p, q, r::Real=2.0)
+    distance(M::ProductManifold, p, q, m::AbstractInverseRetractionMethod=LogarithmicInverseRetraction(), r::Real=2.0)
+
+Compute the distance between `q` and `p` on an [`ProductManifold`](@ref).
+
+First, the componentwise distances are computed. These can be approximated using the
+`norm` of an [`AbstractInverseRetractionMethod`](@ref) `m`.
+Then, the `r`-norm of these elements is computed.
+"""
+
+@doc "$(_doc_distance_prod)"
+function distance(M::ProductManifold, p, q, r::Real = 2.0)
```

Comment:
Shouldn't this use the r-distance on component-like manifolds? In that case I'd actually prefer to have two separate implementations for default `r` and custom `r` instead of a default value here.

Reply:
I do not understand the question nor what you would like to have here, can you elaborate?

Reply:
This _does_ use the component-r-norm thingy.

Reply:
Regarding the first part, let's take `M` to be SE(3), and `p = ArrayPartition([0, 0, 0], I(3))`, `q = ArrayPartition([3, 4, 0], zeros(3, 3))`. Should `distance(M, p, q, Inf)` return 4 or 5? The current code would return 5 but I think 4 would probably be more appropriate?

Reply:
Ah, I se what you mean. It should return 5 and that is correct I think.

It does the usual distance on every component. That is the 2-norm on the R3-part and is 5, the second component is 0.
Only _then_, from these components, the Inf norm is taken.

The Inf can not be passed “recursively”, i.e. Inf does not mean “Do Inf norm on all manifolds where this makes sense” – that to me would be magic; and the `r=` parameter does not exist on all manifolds, so even in code getting 4 generically would be a super-hustle.

TLDR: So with the idea “classical distance on the components, r-norm on the vector” – 5 is correct.

Reply:
And it would be a super huste to test as well. I am already trying to fix/test the corner cases on the existing code and that takes me 2 hours already now. So If you want to have pass-down-inf-magic that would take much longer.

Reply:
OK, 5 is fine, let's just make it clear in docs.